File: CS/PS/복습/progr_lv1/신규아이디추천.py
#17:54 ~ 18:03
def solution(new_id):
    answer = ''
    new_id = list(new_id.lower())
    parse_str = []
    for c in new_id:
        if c.isalnum() or c == '-' or c =='_' or c=='.':
            if parse_str and parse_str[-1] == '.' and c == '.':
                continue
            parse_str.append(c)
    if parse_str and parse_str[0] == '.':
        parse_str.pop(0)
    if parse_str and parse_str[-1] == '.':
        parse_str.pop()
    if not parse_str:
        parse_str.append('a')
    if len(parse_str) >= 16:
        parse_str = parse_str[0:15]
    if parse_str and parse_str[-1] == '.':
        parse_str.pop()
    while len(parse_str) < 3:
        parse_str.append(parse_str[-1])

    
    return ''.join(parse_str)

File: CS/PS/Site/programmers/lv1/신규아이디추천.js
function oneStep(id) {
  let lowerCase = [...id];
  for (let i = 0; i < id.length; i++) {
    if (lowerCase[i] >= "A" && lowerCase[i] <= "Z") {
      lowerCase[i] = lowerCase[i].toLowerCase();
    }
  }

  return lowerCase;
}

function twoStep(id) {
  let arr = [...id];
  arr = arr.filter((letter) => {
    return (
      (letter >= "a" && letter <= "z") ||
      letter === "-" ||
      letter === "_" ||
      letter === "." ||
      (letter >= "0" && letter <= "9")
    );
  });
  return arr;
}
let three;
function threeStep(id) {
  three = id;
  
  if (three.includes("..")) {
    three = three.replace("..", ".");
    threeStep(three);
  } else {
    return three;
  }
  
  return three;
}

function forthStep(id) {
  if (id[0] === ".") {
    id.shift();
  }
  if (id[id.length - 1] === ".") {
    id.pop();
  }
  return id;
}

function fifthStep(id) {
  if (id.length === 0) {
    return "a";
  } else {
    return id;
  }
}

function sixStep(id) {
  if (id.length >= 16) {
    id = id.slice(0, 15);
    if (id[14] === ".") {
      id.pop();
    }
  }
  return id;
}

function lastStep(id) {
  while (id.length < 3) {
    id = [...id, id[id.length-1]];
  }
  return id;
}
function solution(new_id) {
  var answer = "";
  let oneParse = oneStep(new_id);
  let twoParse = twoStep(oneParse);
  let threeParse = threeStep(twoParse.join(""));
  
  let forthParse = forthStep(threeParse.split(""));
  let fifthParse = fifthStep(forthParse);
  let sixParse = sixStep(fifthParse);
  let lastParse = lastStep(sixParse);
  return lastParse.join("");
}

console.log(solution("=.="));
